### PR TITLE
Add Capella types and presets

### PR DESCRIPTION
**Pull request by jimmygchen**
Add Capella types and presets from latest consensus specs
 https://github.com/ethereum/consensus-specs/blob/dev/specs/capella/beacon-chain.md#preset

### CODE REVIEW

File: src/capella/beacon_state.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    pub next_withdrawal_validator_index: ValidatorIndex,
+    pub historical_summaries: List<HistoricalSummary, HISTORICAL_ROOTS_LIMIT>,
+}
+
+#[derive(Default, Debug, SimpleSerialize, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct HistoricalSummary {
+    pub block_summary_root: Root,
+    pub state_summary_root: Root,
+}
```
Comment (ralexstokes) [inline src/capella/beacon_state.rs:73]: I think we can unify this w/ an older utility I had, which we have now ended up shipping anyway...

https://github.com/ralexstokes/ethereum-consensus/blob/main/src/phase0/beacon_state.rs#L44

suggestion: we can rename the existing thing and then import here

File: src/capella/presets/mainnet.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
+pub use crate::altair::mainnet::SYNC_COMMITTEE_SIZE;
```
Comment (ralexstokes) [inline src/capella/presets/mainnet.rs:1]: it looks like this file is not `rustfmt`'d ?

File: src/capella/presets/minimal.rs [hunk 3]
```diff
@@ -0,0 +1,172 @@
+pub use crate::altair::minimal::SYNC_COMMITTEE_SIZE;
```
Comment (ralexstokes) [inline src/capella/presets/minimal.rs:1]: same on `rustfmt` here...

File: src/capella/presets/minimal.rs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    MAX_PROPOSER_SLASHINGS, MAX_VALIDATORS_PER_COMMITTEE, MAX_VOLUNTARY_EXITS,
+    SLOTS_PER_HISTORICAL_ROOT, VALIDATOR_REGISTRY_LIMIT,
+};
+
+pub use capella::*;
+
+pub const MAX_WITHDRAWALS_PER_PAYLOAD: usize = 16;
+pub const MAX_BLS_TO_EXECUTION_CHANGES: usize = 16;
+pub const MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP: usize =16384;
+
```
Comment (ralexstokes) [inline src/capella/presets/minimal.rs:24]: some of these are off; see constants here: https://github.com/ethereum/consensus-specs/blob/dev/presets/minimal/capella.yaml

**Review by ralexstokes — COMMENTED** [review]
this is amazing! thank you :)

left some comments below

I'll make another pass soon to verify we have everything tidy but looks great from first pass

File: src/capella/presets/mainnet.rs (hunk 2 above)
Comment (jimmygchen) [reply src/capella/presets/mainnet.rs:1]: Thanks, should be fixed now! I have "Run rustfmt on Save" enabled in my IDE but for some reasons it didn't work :/ I have rustfmt'd the entire capella directory so this should be good now.

File: src/capella/presets/minimal.rs (hunk 3 above)
Comment (jimmygchen) [reply src/capella/presets/minimal.rs:1]: Done

File: src/capella/presets/minimal.rs (hunk 4 above)
Comment (jimmygchen) [reply src/capella/presets/minimal.rs]: Good catch! fixed now

File: src/capella/beacon_state.rs (hunk 1 above)
Comment (jimmygchen) [reply src/capella/beacon_state.rs]: Thanks, didn't know this exist! I have applied your suggestion.

File: src/capella/withdrawal.rs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+use crate::primitives::{ExecutionAddress, Gwei, ValidatorIndex, WithdrawalIndex};
+use ssz_rs::prelude::*;
+
+#[derive(Default, Debug, Clone, SimpleSerialize, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Withdrawal {
+    pub index: WithdrawalIndex,
+    pub validator_index: ValidatorIndex,
+    pub address: ExecutionAddress,
+    pub amount: Gwei,
```
Comment (ralexstokes) [inline src/capella/withdrawal.rs:10]: going to want `serde` annotations:
```suggestion
    #[serde(with = "crate::serde::as_string")]
    pub index: WithdrawalIndex,
    #[serde(with = "crate::serde::as_string")]
    pub validator_index: ValidatorIndex,
    pub address: ExecutionAddress,
    #[serde(with = "crate::serde::as_string")]
    pub amount: Gwei,
```

File: src/capella/bls_to_execution_change.rs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+use crate::primitives::{BlsPublicKey, BlsSignature, ExecutionAddress, ValidatorIndex};
+use ssz_rs::prelude::*;
+
+#[derive(Default, Debug, Clone, SimpleSerialize, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct BlsToExecutionChange {
+    pub validator_index: ValidatorIndex,
```
Comment (ralexstokes) [inline src/capella/bls_to_execution_change.rs:7]:
```suggestion
    #[serde(with = "crate::serde::as_string")]
    pub validator_index: ValidatorIndex,
```

File: src/capella/execution.rs [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+            extra_data: payload.extra_data.clone(),
+            base_fee_per_gas: payload.base_fee_per_gas.clone(),
+            block_hash: payload.block_hash.clone(),
+            transactions_root,
+            withdrawals_root,
+        })
+    }
+}
+
+pub trait ExecutionEngine<
```
Comment (ralexstokes) [inline src/capella/execution.rs:133]: I haven't thought about how to make the execution bit forward-compatible... I'll come back to this

File: src/capella/beacon_state.rs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    pub finalized_checkpoint: Checkpoint,
+    #[serde(with = "crate::serde::collection_over_string")]
+    pub inactivity_scores: List<u64, VALIDATOR_REGISTRY_LIMIT>,
+    pub current_sync_committee: SyncCommittee<SYNC_COMMITTEE_SIZE>,
+    pub next_sync_committee: SyncCommittee<SYNC_COMMITTEE_SIZE>,
+    pub latest_execution_payload_header: ExecutionPayloadHeader<
+        BYTES_PER_LOGS_BLOOM,
+        MAX_EXTRA_DATA_BYTES,
+        MAX_BYTES_PER_TRANSACTION,
+        MAX_TRANSACTIONS_PER_PAYLOAD,
```
Comment (ralexstokes) [inline src/capella/beacon_state.rs:63]: this is outside of this PR but we can drop these last two generics from the `ExecutionPayloadHeader`

feel free to make this change in a separate PR but ill try to get to this tomorrow

**Review by ralexstokes — COMMENTED** [review]
this is great! I left a few comments

File: src/capella/beacon_state.rs (hunk 8 above)
Comment (ralexstokes) [reply src/capella/beacon_state.rs:63]: working on this here: https://github.com/ralexstokes/ethereum-consensus/pull/175

File: src/capella/execution.rs (hunk 7 above)
Comment (ralexstokes) [reply src/capella/execution.rs:133]: if anything `ExecutionEngine` should just be generic over some type `E` to represent the execution payload but I'll just add an issue to track this

File: src/capella/beacon_state.rs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    >,
+    pub next_withdrawal_index: WithdrawalIndex,
+    pub next_withdrawal_validator_index: ValidatorIndex,
```
Comment (ralexstokes) [inline src/capella/beacon_state.rs:66]:
```suggestion
     #[serde(with = "crate::serde::as_string")]
    pub next_withdrawal_index: WithdrawalIndex,
     #[serde(with = "crate::serde::as_string")]
    pub next_withdrawal_validator_index: ValidatorIndex,
```

Comment (jimmygchen) [reply src/capella/beacon_state.rs]: @ralexstokes thanks for the suggestions and sorry to have missed these! I should've really tested the serialisation / deserialisation output myself - I'll do another pass to make sure I haven't miss anything.

File: src/capella/beacon_state.rs [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    pub latest_execution_payload_header: ExecutionPayloadHeader<
+        BYTES_PER_LOGS_BLOOM,
+        MAX_EXTRA_DATA_BYTES,
+        MAX_BYTES_PER_TRANSACTION,
+        MAX_TRANSACTIONS_PER_PAYLOAD,
+    >,
+     #[serde(with = "crate::serde::as_string")]
+    pub next_withdrawal_index: WithdrawalIndex,
+     #[serde(with = "crate::serde::as_string")]
+    pub next_withdrawal_validator_index: ValidatorIndex,
```
Comment (ralexstokes) [inline src/capella/beacon_state.rs:68]:
```suggestion
    #[serde(with = "crate::serde::as_string")]
    pub next_withdrawal_index: WithdrawalIndex,
    #[serde(with = "crate::serde::as_string")]
    pub next_withdrawal_validator_index: ValidatorIndex,
```

**Pull request merged**